libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1263,7 +1265,41 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				// as require with custom errors is not supported in legacy codegen.
 				auto const* magicType = dynamic_cast<MagicType const*>(arguments[1]->annotation().type);
 				if (magicType && magicType->kind() == MagicType::Kind::Error)
-					solUnimplemented("Require with a custom error is only available using the via-ir pipeline.");
+				{
+					// Make sure that error constructor arguments are evaluated regardless of the require condition
+					auto const& errorConstructorCall = dynamic_cast<FunctionCall const&>(*arguments[1]);
+					errorConstructorCall.expression().accept(*this);
+					std::vector<Type const*> errorConstructorArgumentTypes{};
+					unsigned stackDepth = 0;
+					for (ASTPointer<Expression const> const& errorConstructorArgument: errorConstructorCall.sortedArguments())
+					{
+						errorConstructorArgument->accept(*this);
+						stackDepth += errorConstructorArgument->annotation().type->sizeOnStack();
+						errorConstructorArgumentTypes.push_back(errorConstructorArgument->annotation().type);
+					}
+					// stack: <condition> <arg0> <arg1> ... <argN>
+					// Move condition to the top of the stack
+					utils().moveIntoStack(arguments.at(0)->annotation().type->sizeOnStack(), stackDepth);
```

Comment:
The `copyToStackTop()` could be an interesting proposition, and more likely cheaper than using multiple swaps. Although I guess in that case, we'd have to pop the first instance of condition from the stack regardless of the success/reverting branch. I'll play around with it a bit and see what the gas costs are like.

Also, thanks for the `CompilerUtils::sizeOnStack()`, that's very useful :)

Reply:
OK, so the `DUP` approach is cheaper for errors that have 2 or more arguments; for errors that have one argument it's more expensive, and for errors without arguments it incurs the same cost as the current `SWAP` approach. This makes sense, since the `DUP` approach will always incur an additional `POP` since the initially pushed condition will not be consumed in this case. Most projects typically use custom errors without arguments from what I've seen, e.g. `error Unauthorized(); error InsufficientFunds();` etc. so I don't think it's worthwhile changing the implementation to go in this direction (for legacy). I do agree it's a great idea, but I'm also pretty certain Daniel will not be happy with us spending time on this.

As far as stack too deep is concerned, you are absolutely right that it will happen if you use errors with too many arguments, and the compiler will do exactly what it should be doing - i.e. it will suggest to compile with `--via-ir`.